tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/iterator/FedXZeroLengthPathIteration.java

```diff
@@ -0,0 +1,220 @@
+
+/**
+ * An iteration to evaluated {@link FedXZeroLengthPath}
+ *
+ * @see ZeroLengthPathIteration
+ */
+public class FedXZeroLengthPathIteration extends LookAheadIteration<BindingSet, QueryEvaluationException> {
+
+	/*
+	 * IMPL NOTE:
+	 *
+	 * This is technically almost a 1:1 copy of
+	 * org.eclipse.rdf4j.query.algebra.evaluation.iterator.ZeroLengthPathIteration Reusing or extending
+	 * ZeroLengthPathIteration requires refactoring in its constructor initialization.
+	 *
+	 * The main difference is in keeping track of QueryInfo and statement sources to be used in the precompiled
+	 * statement. Additionally the variable names for anon vars are renamed
+	 */
+
+	private static final Literal OBJECT = SimpleValueFactory.getInstance().createLiteral("object");
+
+	private static final Literal SUBJECT = SimpleValueFactory.getInstance().createLiteral("subject");
+
+	// Note: in contrast to the original zero length path iteration we use "_" instead of "-"
+	// as we need variable names valid in SELECT queries
+
+	private static final String ANON_SUBJECT_VAR = "zero_length_internal_start";
+
+	private static final String ANON_PREDICATE_VAR = "zero_length_internal_pred";
+
+	private static final String ANON_OBJECT_VAR = "zero_length_internal_end";
+
+	private static final String ANON_SEQUENCE_VAR = "zero_length_internal_seq";
+
+	private QueryBindingSet result;
+
+	private final Value subj;
+
+	private final Value obj;
+
+	private final BindingSet bindings;
+
+	private CloseableIteration<BindingSet, QueryEvaluationException> iter;
+
+	private Set<Value> reportedValues;
+
+	private final Var contextVar;
+
+	private final EvaluationStrategy evaluationStrategy;
+
+	private final QueryEvaluationStep precompile;
+
+	private final QueryEvaluationContext context;
+
+	private final BiConsumer<Value, MutableBindingSet> setSubject;
+
+	private final BiConsumer<Value, MutableBindingSet> setObject;
+
+	private final BiConsumer<Value, MutableBindingSet> setContext;
+
+	public FedXZeroLengthPathIteration(EvaluationStrategy evaluationStrategyImpl, Var subjectVar, Var objVar,
+			Value subj, Value obj, Var contextVar, BindingSet bindings, QueryEvaluationContext context,
+			QueryInfo queryInfo, List<StatementSource> statementSources) {
+		this.evaluationStrategy = evaluationStrategyImpl;
+		this.context = context;
+		this.result = new QueryBindingSet(bindings);
+		this.contextVar = contextVar;
+		this.subj = subj;
+		this.obj = obj;
+		this.bindings = bindings;
+		Var startVar = createAnonVar(ANON_SUBJECT_VAR);
+		Var predicate = createAnonVar(ANON_PREDICATE_VAR);
+		Var endVar = createAnonVar(ANON_OBJECT_VAR);
+
+		StatementPattern subjects;
+		if (contextVar != null) {
+			subjects = new StatementPattern(Scope.NAMED_CONTEXTS, startVar, predicate, endVar, contextVar.clone());
+		} else {
+			subjects = new StatementPattern(startVar, predicate, endVar);
+		}
+
+		// specialization for federation: we need to attach statement sources
+		// to the precompiled expr
```

Comment:
this is the only specialization: for the federation we need to send the statement pattern to the identified relevant members. Hence we use a fedx structure which is aware of the sources

Remainder of the code is 1:1